src/main/java/software/amazon/nio/spi/s3/S3WritableByteChannel.java

```diff
@@ -0,0 +1,98 @@
+package software.amazon.nio.spi.s3;
+
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.transfer.s3.S3TransferManager;
+import software.amazon.awssdk.transfer.s3.model.DownloadFileRequest;
+import software.amazon.awssdk.transfer.s3.model.UploadFileRequest;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.nio.file.*;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeoutException;
+
+public class S3WritableByteChannel implements WritableByteChannel {
+    private final S3AsyncClient client;
+    private final S3Path path;
+    private final Path tempFile;
+    private final SeekableByteChannel channel;
+
+    private boolean open;
+
+    public S3WritableByteChannel(S3Path path, S3AsyncClient client, Set<? extends OpenOption> options) throws IOException {
+        Objects.requireNonNull(path);
+        Objects.requireNonNull(client);
+
+        try {
+            S3FileSystemProvider fileSystemProvider = (S3FileSystemProvider) path.getFileSystem().provider();
+            boolean exists = fileSystemProvider.exists(client, path);
+
+            if (exists && options.contains(StandardOpenOption.CREATE_NEW)) {
+                throw new FileAlreadyExistsException("File at path:" + path + " already exists");
+            }
+            if (!exists && !options.contains(StandardOpenOption.CREATE_NEW) && !options.contains(StandardOpenOption.CREATE)) {
+                throw new NoSuchFileException("File at path:" + path + " does exist yet");
+            }
+
+            tempFile = Files.createTempFile("aws-s3-nio-", ".tmp");
+            if (exists) {
+                try (S3TransferManager s3TransferManager = S3TransferManager.builder().s3Client(client).build()) {
+                    s3TransferManager.downloadFile(
+                            DownloadFileRequest.builder()
+                                    .getObjectRequest(GetObjectRequest.builder()
+                                            .bucket(path.bucketName())
+                                            .key(path.getKey())
+                                            .build())
+                                    .destination(tempFile)
+                                    .build()
+                    ).completionFuture().join();
```

Comment:
Can we configure some kind of time out? The danger is that over a slow connection, very large file and/ or other network weirdness this could potentially hang.

Reply:
I also thought about this, but what would be a reasonable value for the timeout? I somewhere read that the S3TransferManager will throw an error if the underlying client receives an timeout, but i can't find the reference anymore and don't know how reliable the statement was.

Reply:
I agree, it is going to be difficult to set a sensible default without knowing how large an object is and what kind of network connection you have. Perhaps the default should be that there is no time out but it would be good to have a version of these methods that allow a caller to provide a maximum time if they would like to.